File: docker/ifcopenshell-worker/ifw/chunking.py
"""Spatial bisection chunk router — Python port of packages/ifc/src/chunking.ts.

The algorithms on both sides MUST stay in lock-step (ticket 09 dual-engine
consistency): same streaming approximation, same guard/bisection semantics.
"""
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple

Point = Tuple[float, float, float]
MAX_SPLITS_PER_STOREY = 32


@dataclass
class ChunkGuard:
    axis: int  # 0|1|2
    limit: float
    upper: bool  # geometry with centroid[axis] >= limit belongs to this chunk


@dataclass
class ChunkTarget:
    index: int
    storey_express_id: Optional[int]
    guards: List[ChunkGuard] = field(default_factory=list)
    triangles: int = 0
    min: List[float] = field(default_factory=lambda: [float("inf")] * 3)
    max: List[float] = field(default_factory=lambda: [float("-inf")] * 3)
    overflowed: bool = False


class ChunkRouter:
    def __init__(self, max_triangles_per_chunk: int) -> None:
        self.max_triangles_per_chunk = max_triangles_per_chunk
        self.chunks: List[ChunkTarget] = []
        self._splits_left: Dict[str, int] = {}
        self._targets: Dict[str, List[ChunkTarget]] = {}

    @staticmethod
    def _key(storey: Optional[int]) -> str:
        return "none" if storey is None else str(storey)

    def _make_target(self, storey: Optional[int], guards: List[ChunkGuard]) -> ChunkTarget:
        target = ChunkTarget(index=len(self.chunks), storey_express_id=storey, guards=guards)
        self.chunks.append(target)
        key = self._key(storey)
        lst = self._targets.setdefault(key, [])
        lst.insert(0, target)  # newest (most constrained) matched first
        return target

    def _locate(self, storey: Optional[int], centroid: Point) -> ChunkTarget:
        lst = self._targets.get(self._key(storey))
        if not lst:
            return self._make_target(storey, [])
        for t in lst:
            if all((centroid[g.axis] >= g.limit) == g.upper for g in t.guards):
                return t
        return lst[-1]  # invariant fallback: root target

    def route(self, storey: Optional[int], centroid: Point, gmin: Point, gmax: Point, tri_count: int) -> ChunkTarget:
        target = self._locate(storey, centroid)
        target.triangles += tri_count
        for a in range(3):
            if gmin[a] < target.min[a]:
                target.min[a] = gmin[a]
            if gmax[a] > target.max[a]:
                target.max[a] = gmax[a]
        if target.triangles <= self.max_triangles_per_chunk or target.overflowed:
            return target

        key = self._key(storey)
        left = self._splits_left.get(key, MAX_SPLITS_PER_STOREY)
        if left <= 0:
            target.overflowed = True  # depth-capped: accept an oversized chunk
            return target
        self._splits_left[key] = left - 1

        # bisect at the midpoint of the target box's longest axis
        axis = 0
        for a in (1, 2):
            if target.max[a] - target.min[a] > target.max[axis] - target.min[axis]:
                axis = a
        limit = (target.min[axis] + target.max[axis]) / 2
        keep_upper = centroid[axis] >= limit
        # the fresh target takes the side the current geometry does NOT belong to
        self._make_target(storey, list(target.guards) + [ChunkGuard(axis, limit, not keep_upper)])
        target.guards.append(ChunkGuard(axis, limit, keep_upper))
        # clip the kept box to the guard so repeated bisections converge
        if keep_upper:
            target.min[axis] = max(target.min[axis], limit)
        else:
            target.max[axis] = min(target.max[axis], limit)
        return target

    def splits_left_for(self, storey: Optional[int]) -> int:
        return self._splits_left.get(self._key(storey), MAX_SPLITS_PER_STOREY)

File: packages/ifc/src/chunking.ts
/**
 * Streaming chunk router (ticket 05).
 *
 * Geometry arrives element-by-element from StreamAllMeshes with no second
 * pass, so spatial bisection is incremental: each storey owns an ordered list
 * of chunk targets; every target is the intersection of half-plane guards
 * inherited from its split ancestors. When a target's accumulated triangle
 * count crosses the budget it is bisected at the midpoint of its longest bbox
 * axis, keeping the side that received the overflowing geometry and moving the
 * other side into a fresh target. Already-streamed geometry stays where it was
 * routed (a streaming approximation); per-chunk bboxes stay exact because they
 * are accumulated from the same centroids' geometry boxes.
 */

export interface ChunkGuard {
  axis: 0 | 1 | 2;
  limit: number;
  /** geometry with centroid[axis] >= limit belongs to this chunk */
  upper: boolean;
}

export interface ChunkTarget {
  /** stable index into the router's chunk list */
  index: number;
  storeyExpressID: number | null;
  guards: ChunkGuard[];
  triangles: number;
  /** world-space AABB of all geometry routed here (metres, Y-up) */
  min: [number, number, number];
  max: [number, number, number];
  /** true once the split depth cap allowed an over-budget chunk to grow */
  overflowed: boolean;
}

const MAX_SPLITS_PER_STOREY = 32;

export class ChunkRouter {
  readonly chunks: ChunkTarget[] = [];
  private splitsLeft = new Map<string, number>();
  /** "storeyExpressID" | "none" -> creation-ordered targets (most specific first) */
  private targets = new Map<string, ChunkTarget[]>();

  constructor(readonly maxTrianglesPerChunk: number) {}

  private static key(storey: number | null): string {
    return storey === null ? "none" : String(storey);
  }

  /** Register a new chunk target and return its stable index. */
  private makeTarget(storey: number | null, guards: ChunkGuard[]): ChunkTarget {
    const target: ChunkTarget = {
      index: this.chunks.length,
      storeyExpressID: storey,
      guards,
      triangles: 0,
      min: [Infinity, Infinity, Infinity],
      max: [-Infinity, -Infinity, -Infinity],
      overflowed: false,
    };
    this.chunks.push(target);
    const key = ChunkRouter.key(storey);
    const list = this.targets.get(key) ?? [];
    list.unshift(target); // newest (most constrained) matched first
    this.targets.set(key, list);
    return target;
  }

  /** First target (creation order) whose guards contain the point. */
  private locate(storey: number | null, centroid: [number, number, number]): ChunkTarget {
    const list = this.targets.get(ChunkRouter.key(storey));
    if (!list || list.length === 0) {
      return this.makeTarget(storey, []);
    }
    for (const t of list) {
      if (t.guards.every((g) => (centroid[g.axis] >= g.limit) === g.upper)) return t;
    }
    // Coverage is an invariant (a split replaces T with T∩half and adds the
    // complement); reaching here means a bug — fall back to the root target.
    return list[list.length - 1];
  }

  /**
   * Route one streamed geometry to a chunk, given its world centroid, world
   * AABB (min/max) and triangle count; splits the receiving target when it
   * exceeds the triangle budget.
   */
  route(storey: number | null, centroid: [number, number, number], gmin: [number, number, number], gmax: [number, number, number], triCount: number): ChunkTarget {
    const target = this.locate(storey, centroid);
    target.triangles += triCount;
    for (let a = 0; a < 3; a++) {
      if (gmin[a] < target.min[a]) target.min[a] = gmin[a];
      if (gmax[a] > target.max[a]) target.max[a] = gmax[a];
    }
    if (target.triangles <= this.maxTrianglesPerChunk || target.overflowed) return target;

    const key = ChunkRouter.key(storey);
    const left = this.splitsLeft.get(key) ?? MAX_SPLITS_PER_STOREY;
    if (left <= 0) {
      target.overflowed = true; // depth-capped: accept an oversized chunk
      return target;
    }
    this.splitsLeft.set(key, left - 1);

    // bisect at the midpoint of the target box's longest axis
    let axis: 0 | 1 | 2 = 0;
    for (let a = 1; a < 3; a++) {
      if (target.max[a] - target.min[a] > target.max[axis] - target.min[axis]) axis = a as 0 | 1 | 2;
    }
    const limit = (target.min[axis] + target.max[axis]) / 2;
    const keepUpper = centroid[axis] >= limit;
    // The fresh target takes the side the current geometry does NOT belong to;
    // the overflowing target keeps its accumulated data and gains its own guard.
    this.makeTarget(storey, [...target.guards, { axis, limit, upper: !keepUpper }]);
    target.guards.push({ axis, limit, upper: keepUpper });
    // Clip the kept box to the guard so repeated bisections converge on the
    // data actually routed here (an unclipped box re-splits at the same plane).
    if (keepUpper) target.min[axis] = Math.max(target.min[axis], limit);
    else target.max[axis] = Math.min(target.max[axis], limit);
    return target;
  }

  /** Number of splits still available for a storey (tests/diagnostics). */
  splitsLeftFor(storey: number | null): number {
    return this.splitsLeft.get(ChunkRouter.key(storey)) ?? MAX_SPLITS_PER_STOREY;
  }
}
